### PR TITLE
metrics: change metadata vector to chunked_fifo

### PR DESCRIPTION
**Pull request by NyaliaLui**
The equivalent PR for upstream scylla is https://github.com/scylladb/seastar/pull/1741 

The vtools test passed CI as well https://github.com/redpanda-data/vtools/pull/1962

Large contiguous allocations may fail as the memory on a shard becomes fragmented over time. This can lead to over allocations on the metadata vector occurring at the call to vector.emplace_back():

```cpp
void impl::update_metrics_if_needed() {
if (_dirty) {
   _metadata = ::seastar::make_shared<metric_metadata>();

   auto mt_ref = ::seastar::make_shared<metric_metadata>();
   auto &mt = *(mt_ref.get());
   mt.reserve(_value_map.size());
   _current_metrics.resize(_value_map.size());
   size_t i = 0;
   for (auto&& mf : _value_map) {
       metric_metadata_vector metrics;
       _current_metrics[i].clear();
       for (auto&& m : mf.second) {
           if (m.second && m.second->is_enabled()) {
               metrics.emplace_back(m.second->info()); // Backtrace shows overalloc occurs here
               _current_metrics[i].emplace_back(m.second->get_function());
           }
       }
       if (!metrics.empty()) {
           mt.emplace_back(metric_family_metadata{mf.second.info(), std::move(metrics)});
           i++;
       }
   }
   _current_metrics.resize(i);
   _metadata = mt_ref;
   _dirty = false;
}
}
```

Therefore, this commit switches the metadata vector with a chunked_fifo.

Changes from force-push `43ea58e`:
- Use a delegating constructor for `chunked_fifo::basic_iterator`
- Split into two commits, one for `chunked_fifo` improvements and another for metrics changes

### CODE REVIEW

**Comment by NyaliaLui** [comment]
Tagging the performance team as reviewers since this PR conceptually has overlap with their work.
CC: @travisdowns @ballard26

File: include/seastar/core/chunked_fifo.hh [hunk 1]
```diff
@@ -210,6 +211,12 @@ private:
 
 };
 
+template <typename T, size_t items_per_chunk>
+template <typename U>
+inline
+chunked_fifo<T, items_per_chunk>::basic_iterator<U>::basic_iterator() noexcept : _chunk(nullptr), _item_index(0) {
```
Comment (rockwotj) [inline include/seastar/core/chunked_fifo.hh:217]: nit: can this be a delegating constructor and pass in a `nullptr` to the explicit constructor that takes a chunk?

File: include/seastar/core/chunked_fifo.hh [hunk 2]
```diff
@@ -142,6 +142,7 @@ private:
         size_t _item_index = 0;
 
     protected:
+        inline basic_iterator() noexcept;
```
Comment (rockwotj) [inline include/seastar/core/chunked_fifo.hh:145]: Why this change? It looks like `iterator` and `const_iterator` both already have default constructors.

**Review by rockwotj — COMMENTED** [review]
nit/clarification in the PR description:

> a std::vector will allocate a large piece of contiguous memory which may become fragmented over time. 

Over time as a seastar shard's memory becomes fragmented, large contiguous allocations may fail.

Sorry if that's a bit pedantic.

File: include/seastar/core/chunked_fifo.hh (hunk 2 above)
Comment (NyaliaLui) [reply include/seastar/core/chunked_fifo.hh:145]: Because of https://github.com/scylladb/seastar/blob/master/src/core/scollectd.cc#L415-L418

Clang was complaining that chunked_fifo did not have a default constructor for that call to create a `shared_ptr<context>()`

Comment (rockwotj) [reply include/seastar/core/chunked_fifo.hh:145]: Ah right and `iterator` and `const_iterator` are marked as `default`.

**Review by rockwotj — APPROVED** [review]
